h2/src/test/org/h2/test/jdbc/TestPreparedStatement.java

```diff
@@ -458,6 +459,28 @@ private void testUUID(Connection conn) throws SQLException {
         stat.execute("drop table test_uuid");
     }
 
+    private void testUUIDAsJavaObject(Connection conn) throws SQLException {
+        String uuidStr = "12345678-1234-4321-8765-123456789012";
+
+        Statement stat = conn.createStatement();
+        stat.execute("create table test_uuid(id uuid primary key)");
+        UUID origUUID = UUID.fromString(uuidStr);
+        PreparedStatement prep = conn.prepareStatement("insert into test_uuid values(?)");
+        prep.setObject(1, origUUID, java.sql.Types.JAVA_OBJECT);
+        prep.execute();
+
+        prep = conn.prepareStatement("select * from test_uuid where id=?");
+        prep.setObject(1, origUUID);
```

Comment:
Ah-- interesting question.  Tracing through things, if the object type is set to java.sql.Types.JAVA_OBJECT, eventually compareTypeSafe( ) in Table.java calls getHigherOrder( ) in Value.java.  This in turn calls getOrder( ) in Value.java, which decides the relative order of the types like this:

...
        case UUID:
            return 42;
        case JAVA_OBJECT:
            return 43;
...

Given that order, JAVA_OBJECTs are considered the higher-order class, so getHigherOrder( ) decides that the operands should be converted to ValueJavaObjects.  That eventually falls through to the default string conversion handling code at the bottom of convertTo( ), which fails because the string version of the UUID contains non-hex chars.


Anyway, the short version of the story is that I think that UUID objects should be considered higher-order objects than JAVA_OBJECTs.  GEOMETRY objects are higher order than JAVA_OBJECTS now, so it seems reasonable to promote UUIDs up above JAVA_OBJECTs, too.

What do you think?